tests/applications.js

```diff
@@ -69,3 +182,352 @@ test('Workflow model from config', t => {
+  const resp = applications.routes.secret(request, response, next);
+
+  response.on('end', () => {
+    t.is(response.statusCode, 200, 'Should be a 200 response');
+
+    return resp.then(res => {
+      t.true(_.includes(res.message, 'update "applications" set "client-secret"'), 'postgres error');
+      t.end();
+    });
+  });
+  response.render();
+});
+
+test.cb.skip('Create new secret - bad referrer', t => {
+  const req = _.cloneDeep(reqObj);
+  req.method = 'POST';
+  req.headers.referrer = '/applications/add';
+  req.session.form.applications.edit.id = 1;
+
+  const request = httpMocks.createRequest(req);
+
+  const response = httpMocks.createResponse({ eventEmitter: EventEmitter });
+  const resp = applications.routes.secret(request, response, next);
+  response.render();
+
+  response.on('end', () => {
+    t.is(response.statusCode, 302, 'Should be a 302 response');
+
+    return resp.then(res => {
+      t.is(res.message, 'Secret can only be changed from the application edit screen', 'should error when bad referrer');
+      t.end();
+    });
+  });
+});
+
+//////////////////////////////
+// Routes - Save application
+//////////////////////////////
+test.cb.skip('Save new app: name required', t => {
+  const req = _.cloneDeep(reqObj);
+  req.method = 'POST';
+  req.session.referrer = '/applications/add';
+  req.url = '/applications/save';
+  req.body = _.cloneDeep(body);
+  req.body['name--text'] = '';
+
+  const request = httpMocks.createRequest(req);
+
+  const response = httpMocks.createResponse({ eventEmitter: EventEmitter });
+  applications.routes.save(request, response);
+
+  response.on('end', () => {
+    t.is(response.statusCode, 302, 'Should be a 302 response');
+    t.is(response._getRedirectUrl(), '/applications/add');
+    t.end();
+  });
+  response.render();
+});
+
+test.cb.skip('Save existing app: name required', t => {
+  const req = _.cloneDeep(reqObj);
+  req.method = 'POST';
+  req.session.referrer = '/applications/123';
+  req.url = '/applications/save';
+  req.body = _.cloneDeep(body);
+  req.body['name--text'] = '';
+
+  const request = httpMocks.createRequest(req);
+
+  const response = httpMocks.createResponse({ eventEmitter: EventEmitter });
+  applications.routes.save(request, response);
+
+  response.on('end', () => {
+    t.is(response.statusCode, 302, 'Should be a 302 response');
+    t.is(response._getRedirectUrl(), '/applications/123');
+    t.end();
+  });
+  response.render();
+});
+
+test.cb.skip('Save new application', t => {
+  const req = _.cloneDeep(reqObj);
+  req.method = 'POST';
+  req.session.referrer = '/applications/add';
+  req.url = '/applications/save';
+  req.body = _.cloneDeep(body);
+
+  const request = httpMocks.createRequest(req);
+
+  const response = httpMocks.createResponse({ eventEmitter: EventEmitter });
+  applications.routes.save(request, response);
+  response.render();
+
+  response.on('end', () => {
+    t.is(response.statusCode, 302, 'Should be a 302 response');
```

Comment:
I'd also like to see a test to ensure that it has been assigned an internal ID (which I guess would go hand-in-hand with ensuring the redirection works)